### ISSUE TITLE
response.code should return nil instead of 0

### ISSUE THREAD

**Issue by Medvezo** [typhoeus/typhoeus#632]
## Problem
In some cases, when request fails, returned code (i.e. HTTP status code) is `0`. Since `0` isn't semantically valid return code, we should return `nil` instead. Do you accept PRs? Happy to contribute one!
## Steps to reproduce

```
request = Typhoeus::Request.new(
      'https://expired.badssl.com/',
      method: :get,
      accept_encoding: 'gzip, deflate',
      http_version: :httpv2,
      maxredirs: 5,
      followlocation: true,
      ssl_verifypeer: true,
      ssl_verifyhost: 2,
      timeout: 10,
      verbose: true,
      tcp_nodelay: true,
)

response = request.run
```

```
puts response.code
=> 0
```

**Comment by andremedeiros** [comment]
We do accept PRs! 

However, I'd love to have some comments from other people about whether this will introduce breakage, as I suspect it's used for error checking and it's expected that `response.code` is a truthy, non-nil value.

**Comment by PericlesTheo** [comment]
I think even though it might not be semantically a valid return code (FYI I wasn't aware of that), it seems that code 0 is something that exists in the wild.

What would the benefit be of changing this now to return `nil` instead of `0`?

**Comment by andremedeiros** [comment]
I think the issue @Medvezo has is that `response.code == 0` indicates that there was potentially a response when in this case the client gave up on the request due to bad SSL. The main question for me is: why don't we throw an exception?

**Comment by Medvezo** [comment]
This [wiki page](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) lists standard and non-standard HTTP status codes.

It is unclear what `HTTP 0` means (+ `response.code` sometimes do return `nil`). There are so many failure modes and and the fact that `response.code` sometimes return `0` and sometimes `nil` makes it harder to rely on that value. If we outright return `nil`, then it will be clear that request has failed and we need to inspect `response.debug_info` or curl's return code to determine the root cause. Relevant stackoverflow [question](https://stackoverflow.com/a/14507670) .

@andremedeiros seems like this library never throws exceptions on failed requests. I always inspect `response.debug_info.to_h[:text]` or `response.return_code` to find out what went wrong

P.S. this reminds me [three-state problem](https://thoughtbot.com/blog/avoid-the-threestate-boolean-problem) on boolean values: `true`, `false`, `nil`

**Comment by Medvezo** [comment]
Another relevant SO question: https://stackoverflow.com/questions/3825581/does-an-http-status-code-of-0-have-any-meaning/12622082#12622082

**Comment by PericlesTheo** [comment]
@Medvezo I agree, I've seen as well so many different failure modes in the wild like that as you mentioned. I've also seen developers using `my_response.code.to_i` so that they are always working with integers and treating 0 and `nil` as the same.

Feel free to submit a PR 👍

**Comment by PericlesTheo** [comment]
As a note, let's make sure we are not breaking the Faraday adapter with this change in any way (in case tests don't pick a breaking change up)
